doc/src/manual/metaprogramming.md

```diff
@@ -11,6 +11,19 @@ code in Julia are represented by Julia data structures, powerful [reflection](ht
 capabilities are available to explore the internals of a program and its types just like any other
 data.
 
+!!! warning
+    Metaprogramming is a powerful tool, but it introduces complexity that can make code more
+    difficult to understand. For example, it can be surprisingly hard to get scope rules
+    correct. Metaprogramming should typically be used only when other approaches such as
+    [higher order functions](@ref man-anonymous-functions) and
+    [closures](https://en.wikipedia.org/wiki/Closure_(computer_programming)) cannot be applied.
+
+    `eval`, `Meta.parse`, and defining new macros should be typically used as a last resort.
```

Comment:
I am not a big fan of mentioning `eval` and `Meta.parse` in the same breath here. There are definitely valid uses for `eval` in top level code, e.g. systematic method or struct definitions. OTOH, I can't really think of any valid usecases of `Meta.parse` outside of implementing your own REPL or other similar IDE tooling.